src/TraitsV2/TraitedMetaclass.class.st

```diff
@@ -0,0 +1,340 @@
+"
+All the traited class metaclasses are instances of myself.
+
+I include all the custom behavior to implement traits.
+I override a number of methods in Metaclass to implement traits.
+
+Also I have the localMethodDict and the traitComposition of the base class.  
+So it is not needed to recompile the methods from TraitedClass. Check #initializeBasicMethods for more details
+"
+Class {
+	#name : #TraitedMetaclass,
+	#superclass : #Metaclass,
+	#instVars : [
+		'localMethods',
+		'composition',
+		'baseLocalMethods',
+		'baseComposition'
+	],
+	#category : #TraitsV2
+}
+
+{ #category : #accessing }
+TraitedMetaclass class >> traitedClassTrait [
+	^ (TaCompositionElement for: TraitedClass)
+]
+
+{ #category : #'accessing method dictionary' }
+TraitedMetaclass >> addAndClassifySelector: selector withMethod: compiledMethod inProtocol: category notifying: requestor [
+	| r |
+	
+	self localMethodDict at: selector put: compiledMethod.
+	
+	r := super
+		addAndClassifySelector: selector
+		withMethod: compiledMethod
+		inProtocol: category
+		notifying: requestor.
+
+	TraitChange addSelector: selector on: self.
+	^ r
+	
+]
+
+{ #category : #'accessing method dictionary' }
+TraitedMetaclass >> addSelector: selector withMethod: compiledMethod notifying: requestor [
+	| r |
+	self localMethodDict at: selector put: compiledMethod.
+	
+	r := super addSelector: selector withMethod: compiledMethod notifying: requestor.
+	TraitChange addSelector: selector on: self.
+	^ r
+]
+
+{ #category : #querying }
+TraitedMetaclass >> allTraits [
+	
+	^ self traitComposition allTraits
+]
+
+{ #category : #accessing }
+TraitedMetaclass >> baseComposition [
+	^ baseComposition
+]
+
+{ #category : #accessing }
+TraitedMetaclass >> baseComposition: anObject [
+	baseComposition := anObject
+]
+
+{ #category : #accessing }
+TraitedMetaclass >> baseLocalMethods [
+	^ baseLocalMethods
+]
+
+{ #category : #accessing }
+TraitedMetaclass >> baseLocalMethods: anObject [
+	baseLocalMethods := anObject
+]
+
+{ #category : #'fileIn/Out' }
+TraitedMetaclass >> definition [
+	"Refer to the comment in ClassDescription|definition."
+
+	^ String streamContents: 
+		[:strm |
+		strm print: self.
+		self hasTraitComposition ifTrue: [
+			strm
+				crtab;
+				nextPutAll: 'uses: ';
+				print: self traitComposition ].
+		
+		(self usesSpecialSlot or: [ Slot showSlotClassDefinition ])
+			ifFalse: [  
+						strm
+							crtab;
+							nextPutAll: 'instanceVariableNames: ';
+							store: self instanceVariablesString]
+			ifTrue: [ 
+						strm
+							crtab;
+							nextPutAll: 'slots: ';
+							nextPutAll: self slotDefinitionString]]
+]
+
+{ #category : #initialization }
+TraitedMetaclass >> emptyMethodDictionary [
+	^ MethodDictionary new: 64.
```

Comment:
It was created bigger, because the TraitedMetaclass instances includes the injected methods from TraitedClass. TraitedClass inserts 29 methods. Following the MethodDIctionary, when it has 29 methods is with 64 positions. That is why